### ISSUE TITLE
No module named 'mainapp'

### ISSUE THREAD

**Issue by pjobson** [axcore/tartube#25]
```
Traceback (most recent call last):
  File "/usr/local/bin/tartube", line 4, in <module>
    __import__('pkg_resources').run_script('tartube==1.1.8', 'tartube')
  File "/home/pjobson/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 666, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/home/pjobson/.local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1469, in run_script
    exec(script_code, namespace, namespace)
  File "/usr/local/lib/python3.6/dist-packages/tartube-1.1.8-py3.6.egg/EGG-INFO/scripts/tartube", line 33, in <module>
ModuleNotFoundError: No module named 'mainapp'
```

Followed the instructions for install: 

```
git clone https://github.com/axcore/tartube.git
cd tartube
python3 setup.py install
```

Throws:

```
/home/pjobson/.local/lib/python3.6/site-packages/setuptools/dist.py:474: UserWarning: Normalizing '1.1.008' to '1.1.8'
  normalized_version,
running install
error: can't create or remove files in install directory
```

Elevated privledges can install it.

```
sudo python3 setup.py install
```

Installs to `/usr/local/bin/tartube`, running as local user or root throws similar errors as shown at the top.

Running Linux Mint 19.2 Tina x86_64.

**Comment by pjobson** [comment]
Oddly if I do:

```
python3 setup.py build
./tartube/tartube
```

It opens the GUI no problem.

**Comment by axcore** [comment]
Sorry, but I don't know how to fix this. I have carefully compared my setup.py and MANIFEST.in against all the documentation and it is still not possible to install Tartube via pip. After spending nearly ten hours on it, I'm giving up.

**Comment by Sargastic** [comment]
I had (on Gentoo Linux) this same problem. Changing all relevant imports to

from tartube import ...

solved it.

**Comment by LennyMcLennington** [comment]
@axcore this might be a bad way of doing things I'm not really a Python expert but it fixed the problem for me when I was trying to install it. This is in [tartube/tartube_debian:31](https://github.com/axcore/tartube/blob/cc6467ad3a716409c2f42c6ccc3038325c681362/tartube/tartube_debian#L31):
```
# Add module directory to path to prevent import issues
import importlib.util
sys.path.append(os.path.abspath(os.path.dirname(importlib.util.find_spec('tartube').origin)))
```
I also had another problem where it was failing to write the settings.json file and adding this in [tartube/mainapp.py:2111](https://github.com/axcore/tartube/blob/cc6467ad3a716409c2f42c6ccc3038325c681362/tartube/mainapp.py#L2111) (and using the tartube_debian file, although I am using arch linux rather than debian) fixed it for me
```
        # If config directory doesn't exist, create it
        config_file_dirname = os.path.dirname(config_file_path)
        if not os.path.exists(config_file_dirname):
            try:
                os.makedirs(config_file_dirname)
            except:
                pass
```

**Comment by axcore** [comment]
That's great, thanks a lot. I'll try it as soon as I have the chance.

**Comment by axcore** [comment]
@LennyMcLennington, I tested your suggestions on a clean installation of Linux Mint 13.3, using:

```
sudo apt install python3-pip
sudo pip3 install tartube
tartube
```

I still get the `ModuleNotFoundError: No module named 'mainapp'` error.

@Sargastic, I tried using `from tartube import ...`, but then Tartube won't execute at all, even from the command line.

I'm not sure where to go from here.

**Comment by LennyMcLennington** [comment]
@axcore if I'm not mistaken, you didn't add `sys.path.append(os.path.abspath(os.path.dirname(importlib.util.find_spec('tartube').origin)))` when you were trying to install it with PyPi, you only added `import importlib.util`

**Comment by axcore** [comment]
@LennyMcLennington It was in [v1.3.091](https://pypi.org/project/tartube/1.3.91/) but not in the current version, [v1.3.092](https://pypi.org/project/tartube/).

**Comment by LennyMcLennington** [comment]
@axcore oh I see, you added it below the `import mainapp`
here it should be like this
```
# Import other modules
import os
import sys
import importlib.util

# Add module directory to path to prevent import issues
spec = importlib.util.find_spec('tartube')
if spec is not None:
    sys.path.append(os.path.abspath(os.path.dirname(spec.origin)))

# Import our modules
import mainapp
```

**Comment by axcore** [comment]
Oops! I tested that on [this release](https://pypi.org/project/tartube/1.3.126/). and it does seem to work. 

Before and after photos (v1.3.091, v1.3.026). The /icons directory does not seem to be installed at all, and I don't know why.

![pypi_test2](https://user-images.githubusercontent.com/37951926/73606278-8e238d80-45a0-11ea-95e1-5c33141528e0.png)
![pypi_test1](https://user-images.githubusercontent.com/37951926/73606281-8fed5100-45a0-11ea-8059-3eb9cf93072b.png)

**Comment by LennyMcLennington** [comment]
Oh I did have this issue before but I forgot, I ended up using PKGBUILD to install it with makepkg. Maybe the best thing to do would be to make deb packages instead of using PyPi so it can install the icons to /usr/local/tartube/icons. I'll see if I can make a deb package now.

**Comment by axcore** [comment]
Someone has already tried setting up a debian package ([here](https://salsa.debian.org/felix-guest/tartube), see [this issue](https://github.com/axcore/tartube/issues/17)), but it's based on v1.0 and has not been updated since.

**Comment by axcore** [comment]
Installation from PyPI is now working. See section 5.3.3 of the [README](https://github.com/axcore/tartube/blob/master/README.rst) for instructions.

**Comment by Vampiloup** [comment]
I don't know if this is still relevant, but I got the same problem today, a few hours after a big system update.
I tried a lot of things, even uninstall / reinstall Python (wich never work, Python devs arent't nice people), changind path, etc.
Nothing worked until I did the Old Thing : I uninstalled Tartube, and then in my ./config folder (I'm on Linux), I deleted the Tartube folder.
Then reinstalled it.
Work well since.

N.B. Really nice soft, by the way.

**Comment by axcore** [comment]
@Vampiloup I'm glad you got it working again (sorry for the inconvenience). From your description I can't guess what might have gone wrong, so I'll just remind any other readers who might stumble on this thread, that uninstalling your system Python can have bad consequences!

**Comment by luk9a** [comment]
After upgrade Debian 12 to 13, Tartube failed to open with error:
```
Traceback (most recent call last):
  File "/home/luke/.local/bin/tartube", line 42, in <module>
    import mainapp
ModuleNotFoundError: No module named 'mainapp'
``` 
Uninstalled and installed again with pipx, it now asks for pygobject, which was solved with libgirepository-2.0-dev dependency.

Thank you

**Comment by axcore** [comment]
> After upgrade Debian 12 to 13, Tartube failed to open with error:

The .deb installer for Tartube v2.5.164 was compiled on Debian 12, so it might not work on Debian 13. 

There are [other methods of installation](https://github.com/axcore/tartube?tab=readme-ov-file#53-installation---linuxbsd), though. For example, you can run Tartube [directly from the source code](https://github.com/axcore/tartube?tab=readme-ov-file#5310-run-without-installing).
